webpack/webpack-workspace-resolve-plugin/src/KnownDescriptionFilePlugin.ts

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import type { Resolver } from 'webpack';
+import type { IPrefixMatch } from '@rushstack/lookup-by-path';
+import type { IResolveContext, WorkspaceLayoutCache } from './WorkspaceLayoutCache';
+
+type ResolveRequest = Parameters<Resolver['hooks']['resolveStep']['call']>[1];
+
+/**
+ * A resolver plugin that optimizes locating the package.json file for a module.
+ */
+export class KnownDescriptionFilePlugin {
+  public readonly source: string;
+  public readonly target: string;
+
+  private readonly _skipForContext: boolean;
+  private readonly _cache: WorkspaceLayoutCache;
+
+  public constructor(cache: WorkspaceLayoutCache, source: string, target: string, skipForContext?: boolean) {
+    this.source = source;
+    this.target = target;
+    this._cache = cache;
+    this._skipForContext = !!skipForContext;
+  }
+
+  public apply(resolver: Resolver): void {
+    if (this._skipForContext && resolver.options.resolveToContext) {
+      return;
+    }
+
+    const target: ReturnType<Resolver['ensureHook']> = resolver.ensureHook(this.target);
+    const { fileSystem } = resolver;
+    function readDescriptionFileAsJson(
+      descriptionFilePath: string,
+      callback: (err: Error | null | undefined, data?: object) => void
+    ): void {
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      fileSystem.readJson!(descriptionFilePath, callback);
+    }
+
+    function readDescriptionFileWithParse(
+      descriptionFilePath: string,
+      callback: (err: Error | null | undefined, data?: object) => void
+    ): void {
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      fileSystem.readFile(descriptionFilePath, (err: Error | null | undefined, data?: string | Buffer) => {
+        if (!data?.length) {
+          return callback(err);
+        }
+        // eslint-disable-next-line @rushstack/no-new-null
+        callback(null, JSON.parse(data.toString()));
+      });
+    }
+
+    const readDescriptionFile: (
+      descriptionFilePath: string,
+      cb: (err: Error | null | undefined, data?: object) => void
+    ) => void = fileSystem.readJson ? readDescriptionFileAsJson : readDescriptionFileWithParse;
```

Comment:
```suggestion
    ) => void = fileSystem.readJson ? fileSystem.readJson: readDescriptionFileWithParse;
```

To avoid the non-null assertion in `readDescriptionFileAsJson`.

Reply:
It'd need to be `fileSystem.readJson.bind(fileSystem)` for safety, but sure.